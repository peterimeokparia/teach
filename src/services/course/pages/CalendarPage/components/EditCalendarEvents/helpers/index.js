export const initializeEventForm = currentEvent => {
    return {
        title : currentEvent ? currentEvent?.event?.title : '',
        location : currentEvent ? currentEvent?.location : '',
        startTime : currentEvent ?  ( currentEvent?.event?.recurringEvent ) ? currentEvent?.event?.rrule?.dtstart : currentEvent?.event?.start : new Date(),
        endTime : currentEvent ? currentEvent?.event?.end : new Date(),
        allDay : currentEvent ? currentEvent?.event?.allDay : false,
        recurringEvent : currentEvent ?  ( currentEvent?.event?.recurringEvent ) : false,
        weekDays : currentEvent ? currentEvent?.event?.weekDays : [],
        endDate : ( currentEvent?.event?.recurringEvent ) ? currentEvent?.event?.rrule?.until : new Date(), 
        frequency : currentEvent ? currentEvent?.event?.rrule?.freq : '',
        interval : currentEvent ? currentEvent?.event?.rrule?.interval : 1,
        duration : currentEvent ? currentEvent?.event?.duration : 1,
        schedulingData : ( currentEvent?.schedulingData?.length > 0 ) ? currentEvent?.schedulingData : [],
        consultation: ( currentEvent?.consultation ) ? currentEvent?.consultation : { coursesInterestedIn: [] }
    };
};