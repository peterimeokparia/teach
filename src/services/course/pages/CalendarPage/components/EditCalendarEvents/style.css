.events-main {
    background-image: url(../../../../images/geometry2.png);
}
.recurring .tooltiptext  {
    visibility: hidden;
    width: 360px;
    background-color: black;
    color: #fff;
    text-align: center;
    border-radius: 6px;
    padding: 5px 0;
    
    /* Position the tooltip */
    position: absolute;
    z-index: 2;

    /* top: -5px;
    left: 125%; */

    /* top: -5px;
    right: 105%; */
}
.recurring:hover .tooltiptext {
    visibility: visible;
    right: 16%;
}
