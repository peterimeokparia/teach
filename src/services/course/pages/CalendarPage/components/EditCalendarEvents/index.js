import { 
connect } from 'react-redux';

import {
saveCalendar,
deleteCalendar } from 'services/course/actions/calendar';

import {
frequencyCollection,
days,
eventEnum } from 'services/course/pages/CalendarPage/helpers';

import {
studentsOption } from 'services/course/pages/CalendarPage/helpers';

import useEditCalendarEventsHook from 'services/course/pages/CalendarPage/hooks/useEditCalendarEventsHook';
import SessionScheduling from 'services/course/pages/CalendarPage/components/TimeLines/SessionScheduling';
import Select from 'react-select';
import DropDown from 'services/course/pages/components/DropDown';
import ToggleButton from 'services/course/pages/components/ToggleButton';
import './style.css';

const EditCalendarEvents = ({
    saveInProgress,  
    onSaveError, 
    error,
    className,
    onSubmit,
    currentUser,
    currentUsers,
    calendar,
    calendars,
    pushNotificationSubscribers,
    deleteCalendar,
    saveCalendar,
    testAdminUsers,
    emailAddresses,
    currentEvent,
    setIsRecurringEvent,
    calendarEventType,
    setIsEditMode,
    setSelectedItemId,
    timeLines,
    children }) => {
    let useEditCalendarEventHookProp = {
        currentUser,
        currentEvent, 
        eventEnum,
        calendarEventType,
        onSubmit,
        setIsRecurringEvent,
        setSelectedItemId,
        setIsEditMode
    };
    let {
        title,
        location,
        allDay,
        recurringEvent,
        weekDays,
        endDate,
        frequency,
        interval,
        freqCollectionData,
        startDateDateTime,
        endDateDateTime,
        startTimeDateTime,
        endTimeDateTime, 
        schedulingData,
        editing,
        inputRef,
        setEventTitle,
        setEventLocation,
        setWeekDays,
        setEndDate,
        setFrequency,
        setInterval,
        setStartDateDateTime,
        setEndDateDateTime,
        setStartTimeDateTime,
        setEndTimeDateTime,
        setSchedulingData,
        reset,
        submit,
        handleRecurringEvent,
        handleAllDayEvent,
        beginEditing
    } = useEditCalendarEventsHook( useEditCalendarEventHookProp );

    if ( saveInProgress ) {
        return <div>...loading</div>;
    } 
    if ( onSaveError ) {
        return <div> { onSaveError.message } </div> ;
    }

const performDelete = () => { // change to saveEvent
    let calendarEvents = calendar?.calendarEvents?.filter(eventToRemove => eventToRemove?.id !== currentEvent?.event?.id );

    saveCalendar( 
    { 
        ...calendar, 
        calendarEvents 
    },
        currentEvent,
        currentUser,
        pushNotificationSubscribers?.filter(subscriber => testAdminUsers.includes( subscriber?.userId ) ),
        emailAddresses
    ); 
};
    
const performDeleteAll = ( calendar ) => { // change??
    deleteCalendar(calendar, currentUser, pushNotificationSubscribers);
};

const cancelEdit = (e) => {
    e.preventDefault();
    reset();
};
 
return editing ? (
    <div className="events-main">       
        <span className="events"> 
            {(calendarEventType === eventEnum.SessionScheduling) &&  
                <SessionScheduling 
                    scheduledStudents={schedulingData}
                    onChange={setSchedulingData}
                    options={studentsOption(currentUsers)}
                /> 
            }
        <span>
        <form className={""} onSubmit={submit}> 
            <span className="row">    
                <input
                    ref={inputRef}
                    type="text" 
                    disabled={saveInProgress} 
                    value={title}
                    onChange={(e) => setEventTitle(e.target.value)}
                    placeholder="Event Title"
                />
            </span>
            <span className="row">    
                <input
                    ref={inputRef}
                    type="text" 
                    disabled={saveInProgress} 
                    value={location} 
                    onChange={(e) => setEventLocation(e.target.value)}
                    placeholder="Event Location"
                />
            </span>
            <span className="row"> 
            <span className='col'>
            {"Start Date"}
               <input
                    ref={inputRef}
                    type="date" 
                    value={startDateDateTime} 
                    onChange={(e) => setStartDateDateTime(e.target.value)}
                    placeholder="Event Start"
                />
            </span>
            <span className="col">
            {"Start Time"}
                <input
                    ref={inputRef}
                    type="time" 
                    disabled={saveInProgress} 
                    value={startTimeDateTime} 
                    onChange={(e) => setStartTimeDateTime(e.target.value)}
                    placeholder="Event Start"
                />
            </span>
            </span> 
            <span className="row"> 
            <span className='col'>
            {"End Date"}
               <input
                    ref={inputRef}
                    type="date" 
                    disabled={saveInProgress} 
                    value={endDateDateTime} 
                    onChange={(e) => setEndDateDateTime(e.target.value)}
                    placeholder="Event End"
                />
            </span>
            <span className="col">
            {"End Time"}
                <input
                    ref={inputRef}
                    type="time" 
                    disabled={saveInProgress} 
                    value={endTimeDateTime} 
                    onChange={(e) => setEndTimeDateTime(e.target.value)}
                    placeholder="Event End"
                />
            </span>
            </span>
            <span className="row"> 
                 <span className="col-sm-3">
                    <label>
                        All Day ?
                    </label>   
                </span>
                <div class="w-10"></div>
                 <span className="col-sm-6"> 
                    <ToggleButton
                        isChecked={allDay}
                        isDisabled={saveInProgress}
                        value={'isAllDay'}
                        onChange={handleAllDayEvent} 
                    />
                 </span>
            </span>
            <span className="row"> 
                 <span className="col-sm-3">
                    <label>
                        Is Recurring ?
                    </label>   
                </span>
                <div class="w-10"></div>
                 <span className="col-sm-6"> 
                    <ToggleButton
                        isChecked={recurringEvent}
                        isDisabled={saveInProgress}
                        value={'isRecurring'}
                        onChange={handleRecurringEvent} 
                    />
                 </span>
            </span>
            {( recurringEvent ) &&  
            <div>
            <span className="row"> 
                    <span className="col-sm-3"> 
                    <label className="recurring">
                        <label className="recurring">
                            Frequency:                                
                        </label> 
                        <label className="recurring">
                        <span className="tooltiptext">
                            How often?
                        </span>
                        </label>
                        </label> 
                    </span>
                    <div class="w-10"></div>
                        <span className="col-sm-6">
                            <DropDown 
                                label={""}
                                key={"_id"}
                                value={"name"}
                                initialValue={frequency}
                                optionCollection={( editing && recurringEvent ) ? freqCollectionData : frequencyCollection }
                                setOptionSelectedValue={setFrequency} 
                            />
                        </span>
                    </span>
                    <label className="recurring">
                    Interval:
                    <span className="tooltiptext">
                        How often do you want to repeat this recurring event per your frequency selection?
                        Default is 1.
                    </span>
                    <input
                        type="number"
                        min="1" 
                        max="100" 
                        disabled={saveInProgress} 
                        value={interval} 
                        onChange={(e) => setInterval(e.target.value)}
                    />
                    </label> 

                    <label className="recurring">
                    Only on these week days:
                    <span className="tooltiptext">
                        Set recurring events only on the weekdays selected.
                    </span>
                    <Select
                        isMulti
                        value={weekDays}
                        onChange={setWeekDays}
                        options={days} 
                    />
                    </label>

                    <label className="recurring">
                    Until:
                    <span className="tooltiptext">
                        Stop this recurring event on the date specified.
                    </span> 
                    <span className="row"> 
                    <span className='col'>
                    {"End Date"}
                    <input
                        ref={inputRef}
                        type="date" 
                        disabled={saveInProgress} 
                        value={endDate} 
                        onChange={(e) => setEndDate(e.target.value)}
                        placeholder="Event End"
                    />
                    </span>
                    <span className="col">
                    {"End Time"}
                    <input
                        ref={inputRef}
                        type="time" 
                        disabled={saveInProgress} 
                        value={endTimeDateTime} 
                        onChange={(e) => setEndTimeDateTime(e.target.value)}
                        placeholder="Event End"
                    />
                    </span>
                    </span>
                    </label>
            </div>
            }  
            { onSaveError && (
                <div className="saveError-message">
                    Error: { onSaveError.message }
                </div>
            )}
             <button onClick={submit} disabled={saveInProgress} > {`Edit`} </button> 
        </form>
        </span>
                 <br></br>
                <form
                    className= {`${className || ''} editing ${error ? 'error' : ''}`} 
                >
                    <span>
                    <input
                        ref={ inputRef }
                        name="reset"
                        type="submit"
                        value={'Reset'}
                        onChange={ cancelEdit }
                    >
                    </input> 
                    </span>         
                </form>
                </span>
                   {error && <div>{error.message}</div>} 
            </div> 
            ) : ( 
                 children(beginEditing, performDelete, performDeleteAll)
                );                         
};

export default connect(null, { saveCalendar, deleteCalendar } )(EditCalendarEvents);