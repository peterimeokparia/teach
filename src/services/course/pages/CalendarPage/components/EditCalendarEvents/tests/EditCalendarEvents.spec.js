import React from 'react';

import { 
Provider, 
useSelector } from "react-redux";

import { 
applyMiddleware } from 'redux'; 

import { 
mockStoreObject } from 'services/course/api';

import EditCalendarEvents  from 'services/course/pages/CalendarPage/components/EditCalendarEvents';
import renderer from 'react-test-renderer';
import thunk from "redux-thunk"; 
import * as reactRedux from 'react-redux';
import configureStore from 'redux-mock-store';

  const mockStore = configureStore([]);

describe('EditCalendarEvents', () =>  {  

  let  store = mockStore(mockStoreObject, applyMiddleware(thunk));

  let setSelectedItemId = jest.fn();
  let  setIsEditMode = jest.fn(true)

  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');
  let children = jest.fn(({ children }) => (
    <div>{children}</div> ));

  beforeEach(() => {
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
    jest.spyOn(Date.prototype, 'toLocaleTimeString')
      .mockImplementation(() => '12:00:00');
    jest.spyOn(Date.prototype, 'toLocaleDateString')
      .mockImplementation(() => 'AM 1/01/2021');

  
  });

  it('renders as expected', async () => {
    const dispatch = jest.fn();
    useDispatchMock.mockReturnValue(dispatch)

    const tree = renderer.create(
      <Provider store={store}>
        <EditCalendarEvents 
            setSelectedItemId={() =>  setSelectedItemId("001")}
            setIsEditMode={setIsEditMode}
        >
          {
             ( beginEditing, performDelete, performDeleteAll ) => {
                beginEditing() 
             } 
           }
        </EditCalendarEvents>
    </Provider>
    );
      expect(tree.toJSON()).toMatchSnapshot();
  })

  it('renders children as expected', async () => {
    const tree = renderer.create(
      <Provider store={store}>
        <EditCalendarEvents>
          {
             ( beginEditing, performDelete, performDeleteAll ) => {
                   return(<div>
                    <form>
                      <input />
                    </form>
                  </div>)
             } 
           }
        </EditCalendarEvents>
    </Provider>
    );
      expect(tree.toJSON()).toMatchSnapshot();
  })
});


//https://dev.to/d_ir/verifying-children-passed-to-react-mock-components-2mf9


