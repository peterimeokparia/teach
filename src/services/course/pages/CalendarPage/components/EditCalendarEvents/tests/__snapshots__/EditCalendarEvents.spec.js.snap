// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditCalendarEvents renders as expected 1`] = `
<div
  className="events-main"
>
  <span
    className="events"
  >
    <span>
      <form
        className=""
        onSubmit={[Function]}
      >
        <span
          className="row"
        >
          <input
            onChange={[Function]}
            placeholder="Event Title"
            type="text"
            value=""
          />
        </span>
        <span
          className="row"
        >
          <input
            onChange={[Function]}
            placeholder="Event Location"
            type="text"
            value=""
          />
        </span>
        <span
          className="row"
        >
          <span
            className="col"
          >
            Start Date
            <input
              onChange={[Function]}
              placeholder="Event Start"
              type="date"
            />
          </span>
          <span
            className="col"
          >
            Start Time
            <input
              onChange={[Function]}
              placeholder="Event Start"
              type="time"
            />
          </span>
        </span>
        <span
          className="row"
        >
          <span
            className="col"
          >
            End Date
            <input
              onChange={[Function]}
              placeholder="Event End"
              type="date"
            />
          </span>
          <span
            className="col"
          >
            End Time
            <input
              onChange={[Function]}
              placeholder="Event End"
              type="time"
            />
          </span>
        </span>
        <span
          className="row"
        >
          <span
            className="col-sm-3"
          >
            <label>
              All Day ?
            </label>
          </span>
          <div
            class="w-10"
          />
          <span
            className="col-sm-6"
          >
            <div>
              <label
                className="switch"
              >
                <input
                  checked={false}
                  onChange={[Function]}
                  type="checkbox"
                  value="isAllDay"
                />
                <span
                  class="slider round"
                />
              </label>
            </div>
          </span>
        </span>
        <span
          className="row"
        >
          <span
            className="col-sm-3"
          >
            <label>
              Is Recurring ?
            </label>
          </span>
          <div
            class="w-10"
          />
          <span
            className="col-sm-6"
          >
            <div>
              <label
                className="switch"
              >
                <input
                  checked={false}
                  onChange={[Function]}
                  type="checkbox"
                  value="isRecurring"
                />
                <span
                  class="slider round"
                />
              </label>
            </div>
          </span>
        </span>
        <button
          onClick={[Function]}
        >
           
          Edit
           
        </button>
      </form>
    </span>
    <br />
    <form
      className=" editing "
    >
      <span>
        <input
          name="reset"
          onChange={[Function]}
          type="submit"
          value="Reset"
        />
      </span>
    </form>
  </span>
</div>
`;

exports[`EditCalendarEvents renders children as expected 1`] = `
<div>
  <form>
    <input />
  </form>
</div>
`;
